### ISSUE TITLE
OmniCore Controller Support | GoFa CRB1500 ROS support

### ISSUE THREAD

**Issue by DasSchilli** [ros-industrial/abb_robot_driver#33]
Hello,

Can I use the abb_robot_driver for this controller? On the controller runs RobotWare 7.3.1. I have already installed all abb pkgs and these pkg runs successfully. 

In my project for my master study I need only the /joint_states topic, as fast as possible. So I decided to use the EGM Position Stream. Before I test it on the real controller, I want to setup it at RobotStudio. I have already a station with a virtual controller and EGM Position Stream in the Rapid Code. I can also ping vice versa between Windows and Linux. But I get no data between these two machines, the Linux machine is always waiting, but the Rapid code runs successfully. I use the sample code from the application guide. Furthermore I also checked with command "lsof -i" on Ubuntu side, and verified ROS node is running with port 6511 open.

 If EGM is not support, can I use the RWS to get the Joint States, if yes what is the maximum polling rate ?

Linux Machine:

- Ubuntu 20.04 LTS
- ROS Noetic

**Comment by gavanderhoorn** [comment]
I know @jontje has done some tests, but for now please see the readme of `abb_egm_hardware_interface` ([here](https://github.com/ros-industrial/abb_robot_driver/tree/32eb0f48abfd3358cb5a4f3f043f1306a9811b24/abb_egm_hardware_interface#requirements)):

> ### Requirements
> 
>  - `RobotWare` version `6.07.01` or higher (less than `7.0`).

**Comment by gavanderhoorn** [comment]
If you'd like to help: contact your local ABB office and tell them you'd like OmniCore support in their ROS `abb_robot_driver` packages (link them to this repository).

That would show them clearly there is interest.

**Comment by gavanderhoorn** [comment]
As to CRB 15000 support: there is a preliminary package available over at [ros-industrial/abb_experimental](https://github.com/ros-industrial/abb_experimental).

**Comment by DasSchilli** [comment]
Thank you for the quick reply. 

Ok, yes I found these requirements, but I also found the abb_libegm, there is only the requirement RobotWare 6.07.01 or higher. And I found a abb_libegm_samples.zip with a examples of setting up an EGM Server inside a rosnode. But I'm not sure if the current abb_libegm works with RW7.

Yes I will contact my local ABB Office. The CRB 15000 support works already on my machine.

**Comment by gavanderhoorn** [comment]
I would suggest to assume something is not supported unless it is explicitly mentioned.

Especially when migrating to a new controller + RW generation.

**Comment by DasSchilli** [comment]
Is the support planned until this year? because I have to search for a solution for my project.

**Comment by gavanderhoorn** [comment]
Perhaps @jontje can answer this one.

**Comment by gavanderhoorn** [comment]
In any case: please forward your request to your local ABB office as well. Without customer demand it's difficult to allocate resources.

**Comment by DasSchilli** [comment]
FYI, I tried the abb_libegm_example with a virtual (RW 7.3.1) controller in RobotStudio and a Linux PC with ROS Noetic. The result is the EGM communication between these two machines works. The next step is to read the joint values and publish it as a rostopic /joint_states.

**Comment by gavanderhoorn** [comment]
I'm happy for you it works, but to get full support for the driver work is needed. So if you haven't done so yet, please reach out to your local ABB branch office.

**Comment by DasSchilli** [comment]
Yes I read that the egm_hardware_interface is provided by the abb_rws_service_provider package, is required during initialization of the hardware interface. (for getting full support of the abb_robot_driver)

I have already contact my contact person at ABB for this project.

**Comment by DasSchilli** [comment]
> I'm not sure I follow. `abb_egm_hardware_interface` is a stand-alone package. It does use the RWS packages right now, but it's not part of it.
> 
I found ([here](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_egm_hardware_interface#overview))the note:

**Comment by gavanderhoorn** [comment]
No, RWS is indeed a hard requirement. It's used to retrieve configuration information from the controller and to coordinate during the EGM session.

**Comment by jontje** [comment]
> Is the support planned until this year?

Adding support of RWS 2.0 to [abb_librws](https://github.com/ros-industrial/abb_librws) is unlikely to happen this year.

Anyway, it is possible to use the [abb_egm_hardware_interface](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_egm_hardware_interface) ROS package with OmniCore controllers, if you _manually specify a description of the ABB robot controller's system_.

For example, this [abb_rc_description_gofa_example.zip](https://github.com/ros-industrial/abb_robot_driver/files/7267729/abb_rc_description_gofa_example.zip) file contains a ROS package with a simple example node, which provides the minimal required robot controller description for a GoFa setup.

To use this with the second example in [abb_robot_bringup_examples](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_robot_bringup_examples) you only need to replace these [lines](https://github.com/ros-industrial/abb_robot_driver/blob/32eb0f48abfd3358cb5a4f3f043f1306a9811b24/abb_robot_bringup_examples/launch/ex2_rws_and_egm_6axis_robot.launch#L13-L21) with the node from the example package.

Please note that I have only tested this briefly against a RobotStudio simulation.

**Comment by gavanderhoorn** [comment]
Ah, right. I'd forgotten we actually designed this to use a service *exactly* for this kind of use-case (ie: where RWS could not, or should not, be used to describe the robot geometry and software options).

Nice @jontje :+1:

**Comment by jontje** [comment]
> But IIUC, you still can't use RWS to _coordinate_ the IRC5 side of things.

Yes, that is true (I assume you meant OmniCore instead of IRC5 😉).

**Comment by gavanderhoorn** [comment]
Right, of course. I actually was thinking "abb controller" in general, but it's specifically OmniCore of course here.

**Comment by DasSchilli** [comment]
> > Is the support planned until this year?
> 
> Adding support of RWS 2.0 to [abb_librws](https://github.com/ros-industrial/abb_librws) is unlikely to happen this year.
> 
> Anyway, it is possible to use the [abb_egm_hardware_interface](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_egm_hardware_interface) ROS package with OmniCore controllers, if you _manually specify a description of the ABB robot controller's system_.
> 
> For example, this [abb_rc_description_gofa_example.zip](https://github.com/ros-industrial/abb_robot_driver/files/7267729/abb_rc_description_gofa_example.zip) file contains a ROS package with a simple example node, which provides the minimal required robot controller description for a GoFa system.
> 
> To use this with the second example in [abb_robot_bringup_examples](https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_robot_bringup_examples) you only need to replace these [lines](https://github.com/ros-industrial/abb_robot_driver/blob/32eb0f48abfd3358cb5a4f3f043f1306a9811b24/abb_robot_bringup_examples/launch/ex2_rws_and_egm_6axis_robot.launch#L13-L21) with the node from the example package.
> 
> Please note that I have only tested this briefly against a RobotStudio simulation.

Thank you for your support, in the next two weeks I can test it on a real controller.

**Comment by DasSchilli** [comment]
The communication with the OmniCore controller + GoFa via EGM is working. I have adjusted the ABB Rapid Code in the real controller. In the abb_libegm_samples.zip in the while loop a EGMSetupUC has been performed, this leads partly to crashes after a certain time (only on the real controller). I do EGMSetupUC at the beginning (with a WaitTime), now it works fine.

`
MODULE Module1
    ! Identifier for the EGM correction.
    LOCAL VAR egmident egm_id;
    
    PROC main()
        ! Register an EGM id.
        EGMGetId egm_id;

	WaitTime 3;
        ! Setup the EGM communication.
        EGMSetupUC ROB_1, egm_id, "default", "ROB_1", \Joint;

	WaitTime 3;
	EGMStreamStartegm_id\SampleRate:=4;

        WHILE TRUE DO
		MoveAbs][[12.3881,-20.4178,28.5566,-2.80517,41.8339,47.7017],[9E+09,9E+09,9E+1
		MoveAbs][[12.3891,-20.418,39.5557,-2.80544,41.8341,47.7017],[9E+09,9E+09,9E+0:
		MoveAbs][[72.2425,0.929214,22.7271,-4.44463,42.1747,47.7019],[9E+09,9E+09,9E+1
		MoveAbs][[28.7876,22.5377,-0.0826544,-4.55474,42.2717,47.7019],[9E+09,9E+09,91
		MoveAbs][[-47.2852,-27.6126,21.1969,-4.3998,58.5481,47.70341.[9E+09,9E+09,9E+1
		MoveAbs][[6.34636,-23.2613,29.0606,-5.14666,71.33,47.5774],[9E+09,9E+09,9E+09
        ENDWHILE
        
	! Release the EGM id.
        EGMReset egm_id;

    ENDPROC
ENDMODULE
`

**Comment by gavanderhoorn** [comment]
thanks for reporting your success @DasSchilli :+1:

**Comment by DasSchilli** [comment]
Hello,

Sorry for my late reply. Were you able to solve your problem?

**Comment by Yadunund** [comment]
@DasSchilli 

Thanks for the reply! It's nice to hear from you. 

Unfortunately I haven't solved the issue yet.

I was able to customize the station setup to use the Omnicore controller with the irb1200. But I've also set up a station with an irb1300 which only ships with the omnicore and the problem persists. Any pointers would be much appreciated.
Alternatively if you have a pack&go file of a setup with omnicore that you can share, could you send it to me at yadunund@openrobotics.org? 🙇

**Comment by DasSchilli** [comment]
Ok, I will send you my files.

**Comment by DasSchilli** [comment]
sorry your E-Mail is wrong.

**Comment by Yadunund** [comment]
> sorry your E-Mail is wrong.

Hmm I don't see any problems with that address. It should work. Alternatively yadunund@gmail.com works too. Many thanks once again.

**Comment by pucciland95** [comment]
Hi guys, 

For future reference, we are currently working to support the robots Gofa and Yumi Single Arm with ROS1. I leave the link here in case you want more information: https://github.com/MerlinLaboratory/abb_wrapper

The repo already works with both robots and supports ros_control and Moveit! 

Cheers

**Comment by ghx0x0** [comment]
> > sorry your E-Mail is wrong.
> 
> Hmm I don't see any problems with that address. It should work. Alternatively [yadunund@gmail.com](mailto:yadunund@gmail.com) works too. Many thanks once again.

Hi, Yadunund, I met the same problem with yours, "Not possible to open the UdpUc device" on a irb1300_7_1400 arm. Can you give me a help about how to solve this problem with a OmniCore controller? Thanks.

**Comment by Yadunund** [comment]
> > > sorry your E-Mail is wrong.
> > 
> > 
> > Hmm I don't see any problems with that address. It should work. Alternatively [yadunund@gmail.com](mailto:yadunund@gmail.com) works too. Many thanks once again.
> 
> Hi, @Yadunund , I met the same problem with yours, "Not possible to open the UdpUc device" on a irb1300_7_1400 arm. Can you give me a help about how to solve this problem with a OmniCore controller? Thanks.

I wasn't able to get the connection to work in RobotStudio but I last tried with version 2022.2. Not sure if has been fixed in the latest RobotStudio. But I had access to a real robot and was able to successfully control the robot over EGM (after bypassing RWS and using a static description of the robot).

**Comment by ghx0x0** [comment]
> OK, I got it. Sounds like an issue between Omnicore and RobotStudio. I am using RobotStudio-2023 with a RobotWare 7.10.0, this issue is still not solved yet. 
Thanks for your help, I will try it using a real robot.

**Comment by pucciland95** [comment]
Hi guys, 

Unfortunately, this is not a bug of ABB.
Some time ago, I spoke with some guys of the R&D from ABB about this problem and it turns out that (at least when simulating with RobotStudio) this behaviour is expected.
RobotStudio does not let you communicate with external devices through UDP except for the localhost (127.0.0.1) when using Omnicore Controller. Apparently, this is related to CyberSecurity problems.

Clearly, this does not apply to RWS since it communicates with TCP/IP and then works both with the real robot and the one simulated in RobotStudio. 
To recap this is what works:

|           | Real       | Sim |
| --------| ----------- | -------------------- |
| EGM  | Works   | Not Working   |
| RWS  | Works   |  Working          |

I hope this will change in the future,

Cheers

**Comment by gavanderhoorn** [comment]
I'm curious whether what's discussed in https://github.com/ros-industrial/abb_librws/issues/81 could help here.

**Comment by pucciland95** [comment]
Unlucky no, 

What you are referencing is about RWS communication. 
Whitelisting the ip of the machine where ROS is allow you to use RWS. 

EGM unluckily still **does not work** even after the whitelisting.

**Comment by Yadunund** [comment]
> RobotStudio does not let you communicate with external devices through UDP except for the localhost (127.0.0.1) when using Omnicore Controller. Apparently, this is related to CyberSecurity problems.

That is an interesting insight. I wonder if some iptable setup which forwards traffic from localhost to the network interface could overcome this limitation.

**Comment by pucciland95** [comment]
I tried that with no luck (probably cause I know nothing about that). 

I also tried to do a proxy server that sends and receive UDP messages back and forth but I only manage to take data from RobotStudio.

It is not worth the effort in my opinion.

**Comment by Panda12580** [comment]
> 感谢您的快速回复。
> 
> 好的，是的，我找到了这些要求，但我还找到了 abb_libegm，它只要求 RobotWare 6.07.01 或更高版本。我还找到了一个 abb_libegm_samples.zip，其中包含在 rosnode 内设置 EGM 服务器的示例。但我不确定当前的 abb_libegm 是否适用于 RW7。
> 
> 是的，我会联系当地的 ABB 办事处。CRB 15000 支持已在我的计算机上运行。

hallo, I also use the CRB15000 series, but I don't seem to see EGM plug-in on the instructor. Do you have it? How did you connect to ros? I am a novice and I don't seem to understand how to achieve these things, I hope you can help me.
